### ISSUE TITLE
BigInteger Shrink appears to be broken

### ISSUE THREAD

**Issue by jbgi** [functionaljava/functionaljava#276]
Current implementation: https://github.com/functionaljava/functionaljava/blob/3716e739d01f3ef03216e02b16174d6a0ec9b6b3/quickcheck/src/main/java/fj/test/Shrink.java#L673

as per original comment by @bakeemawaytoys in #274:

There is a call to System.arraycopy in the implementation that throws an exception because the source array is byte[] and the destination array is Byte[]. Once fixed, BigInteger shrink and doesn't appear to shrink correctly.

For example
```java
Shrink.shrinkLong.shrink(1000L).take(100).toList();
//Produces List(0,0,500,750,875,938,969,985,993,997,999)
```
versus
```java
Shrink.shrinkBigInteger.shrink(BigInteger.valueOf(1000L)).take(100).toList();
//Produces List(0,0,0,512,0,0,0,512,2,2,2,514)
```
Modifying Shrink.shrinkBigInteger to use the same logic as Shrink.shrinkLong appears to work correctly.

@tonymorris / @mperry / @mrbackend: Would you confirm this is a bug? Is the suggested fix correct?

**Comment by mperry** [comment]
Scalaprops looks like it has a similar implementation, though I have not tried running it:

https://github.com/scalaprops/scalaprops/blob/master/core/src/main/scala/scalaprops/Shrink.scala

Using ghci gives the same answer for both Int64, Integer and Int types:

```
Prelude> :m Test.QuickCheck
Prelude Test.QuickCheck> shrink (1000 :: Data.Int.Int64)
[0,500,750,875,938,969,985,993,997,999]
Prelude Test.QuickCheck>  shrink (1000 :: Integer)
[0,500,750,875,938,969,985,993,997,999]
Prelude Test.QuickCheck> shrink (1000 :: Int)
[0,500,750,875,938,969,985,993,997,999]
```

Not sure if this helps, but it looks like a bug to me.

I thought `Shrink<A>` produced a stream of simpler values so that we could check if they also fail the test.  `shrinkBigInteger` does not seem to be doing that.  I wonder what the history of the development of Scalaprops and this implementation is.

**Comment by mperry** [comment]
Frege copies the Haskell implementation:

https://hackage.haskell.org/package/QuickCheck-2.8.2/docs/src/Test-QuickCheck-Arbitrary.html
https://github.com/Frege/frege/blob/master/frege/test/QuickCheckArbitrary.fr

**Comment by mperry** [comment]
Fixed in #277.
